src/mapgen/cavegen.cpp

```diff
@@ -96,13 +98,32 @@ void CavesNoiseIntersection::generateCaves(MMVManip *vm,
 		u16 base_filler = depth_top + biome->depth_filler;
 		u16 depth_riverbed = biome->depth_riverbed;
 		u16 nplaced = 0;
+
+		s16 nextBiomeY = biome->min_pos.Y;
+
 		// Don't excavate the overgenerated stone at nmax.Y + 1,
 		// this creates a 'roof' over the tunnel, preventing light in
 		// tunnels at mapchunk borders when generating mapchunks upwards.
 		// This 'roof' is removed when the mapchunk above is generated.
 		for (s16 y = nmax.Y; y >= nmin.Y - 1; y--,
 				index3d -= m_ystride,
 				VoxelArea::add_y(em, vi, -1)) {
+			// We need this check to make sure that biomes don't generate too far down
+			if (y < nextBiomeY) {
+				biome = m_bmgn->getBiomeAtIndex(index2d, v3s16(x, y, z));
+				// If the new biome goes to the top of the next biome this will be used
+				nextBiomeY = biome->min_pos.Y;
+
+				// Otherwise we need to make sure that we don't miss a biome because it is ontop of another one
+				for (size_t i = 1; i < m_bmgr->getNumObjects(); i++) {
```

Comment:
This code does not filter against humidity and heat. Same for `MapgenBasic::generateBiomes`.
Idea: Change `getBiomeAtIndex` to `getBiomeAtIndex(size_t index, v3s16 pos, float *min_y = nullptr)`. This way you would be able to "return" `min_y` from `calcBiomeFromNoise`.

Reply:
I just pushed a new commit that should fix this. I did what you recommended and added a field to getBiomeAtIndex called min_y.

Only thing that is a bit strange that currently happens is that if you are in the vertical blending zone of two biomes and you call getBiomeAtIndex you might get a min_y above the y you queried for. This is because I can't calculate the height that blending will happen due to how the RNG is done.